### PR TITLE
feat(hcl2cdk): support variable types and validations

### LINKED ISSUE #2423: CDKTF init fails when importing existing project
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### cdktf & Language Versions

CDKTF v0.14.3

### Affected Resource(s)

All

### Debug Output

```
Error: Error: HCL-JSON does not conform to schema. This is not expected, please file a bug under https://cdk.tf/bugs/new/convert
Please include this information:
[{"code":"invalid_union","unionErrors":[{"errors":[{"code":"invalid_type","expected":"object","received":"string","path":["variable","environment",0,"validation",0,"condition"],"message":"Expected object, received string"},{"code":"invalid_type","expected":"object","received":"string","path":["variable","environment",0,"validation",0,"error_message"],"message":"Expected object, received string"}]},{"errors":[{"code":"invalid_type","expected":"undefined","received":"array","path":["variable","environment",0,"validation"],"message":"Expected undefined, received array"}]}],"path":["variable","environment",0,"validation"],"message":"Invalid input"}]
    at getParsedHcl (/Users/username/.nvm/versions/node/v18.12.0/lib/node_modules/cdktf-cli/node_modules/@cdktf/hcl2cdk/lib/index.js:62:15)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async parseProviderRequirements (/Users/username/.nvm/versions/node/v18.12.0/lib/node_modules/cdktf-cli/node_modules/@cdktf/hcl2cdk/lib/index.js:71:18)
    at async $ot (/Users/username/.nvm/versions/node/v18.12.0/lib/node_modules/cdktf-cli/bundle/bin/cmds/handlers.js:492:1751)
    at async Object.zFr (/Users/username/.nvm/versions/node/v18.12.0/lib/node_modules/cdktf-cli/bundle/bin/cmds/handlers.js:492:26038)
    at async fp.handleCommand (/Users/username/.nvm/versions/node/v18.12.0/lib/node_modules/cdktf-cli/bundle/bin/cdktf.js:53:31390)
    at async B.handler (/Users/username/.nvm/versions/node/v18.12.0/lib/node_modules/cdktf-cli/bundle/bin/cdktf.js:53:29879)
Collecting Debug Information...
/Users/username/.nvm/versions/node/v18.12.0/lib/node_modules/cdktf-cli/node_modules/yoga-layout-prebuilt/yoga-layout/build/Release/nbind.js:53
        throw ex;
        ^

Error: Error: HCL-JSON does not conform to schema. This is not expected, please file a bug under https://cdk.tf/bugs/new/convert
Please include this information:
[{"code":"invalid_union","unionErrors":[{"errors":[{"code":"invalid_type","expected":"object","received":"string","path":["variable","environment",0,"validation",0,"condition"],"message":"Expected object, received string"},{"code":"invalid_type","expected":"object","received":"string","path":["variable","environment",0,"validation",0,"error_message"],"message":"Expected object, received string"}]},{"errors":[{"code":"invalid_type","expected":"undefined","received":"array","path":["variable","environment",0,"validation"],"message":"Expected undefined, received array"}]}],"path":["variable","environment",0,"validation"],"message":"Invalid input"}]
    at getParsedHcl (/Users/username/.nvm/versions/node/v18.12.0/lib/node_modules/cdktf-cli/node_modules/@cdktf/hcl2cdk/lib/index.js:62:15)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async parseProviderRequirements (/Users/username/.nvm/versions/node/v18.12.0/lib/node_modules/cdktf-cli/node_modules/@cdktf/hcl2cdk/lib/index.js:71:18)
    at async $ot (/Users/username/.nvm/versions/node/v18.12.0/lib/node_modules/cdktf-cli/bundle/bin/cmds/handlers.js:492:1751)
    at async Object.zFr (/Users/username/.nvm/versions/node/v18.12.0/lib/node_modules/cdktf-cli/bundle/bin/cmds/handlers.js:492:26038)
    at async fp.handleCommand (/Users/username/.nvm/versions/node/v18.12.0/lib/node_modules/cdktf-cli/bundle/bin/cdktf.js:53:31390)
    at async B.handler (/Users/username/.nvm/versions/node/v18.12.0/lib/node_modules/cdktf-cli/bundle/bin/cdktf.js:53:29879)

Node.js v18.12.0
```

### Expected Behavior

Create CDK project based on existing project.

### Actual Behavior

Application crashed.

### Steps to Reproduce

Run cdktf init command and pass in existing project path.

### Important Factoids

Real-world application infrastructure.

### References

None.

### LINKED ISSUE #2764: Count should be handled with API in convert

### PR DESCRIPTION
**Pull request by DanielMSchmidt**
Closes #2423

### CODE REVIEW

File: packages/@cdktf/hcl2json/lib/syntax-tree.ts [hunk 1]
```diff
@@ -192,6 +197,11 @@ export type BinaryOpExpression = CommonExpressionAst & {
   meta: BinaryOpExpressionMeta;
 };
 
+export type ObjectExpression = CommonExpressionAst & {
+  type: "object";
+  meta: ObjectExpressionMeta; // Doesn't have any special meta attributes
```
Comment (ansgarm) [inline packages/@cdktf/hcl2json/lib/syntax-tree.ts:202]: What does this comment mean? It seems like it has `items` – remove the comment?

**Review by ansgarm — COMMENTED** [review]
only a few very small things 👀

File: packages/@cdktf/hcl2cdk/lib/types.ts [hunk 2]
```diff
@@ -19,6 +19,7 @@ export type ProgramScope = {
   >;
   // Temporary flag to indicate if we need to import the cdktf library to access the token class
   hasTokenBasedTypeCoercion: boolean;
+  nodeIds: string[]; // temporarily added until replaced
```
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply packages/@cdktf/hcl2cdk/lib/types.ts:22]: We only need nodeIds because it made regex based parsing easier (knowing what you are looking for). We now have a proper expression parser so we need to replace the function searching for references with it.

File: packages/@cdktf/hcl2cdk/test/__snapshots__/variables.test.ts.snap [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    super(scope, name);
+    /*Terraform Variables are not always the best fit for getting inputs in the context of Terraform CDK.
+    You can read more about this at https://cdk.tf/variables*/
+    new local.provider.LocalProvider(this, \\"local\\", {});
+    const imageId = new cdktf.TerraformVariable(this, \\"image_id\\", {
+      description: \\"The id of the machine image (AMI) to use for the server.\\",
+      validation: [
+        {
+          condition:
+            '\${length(var.image_id) > 4 && substr(var.image_id, 0, 4) == \\"ami-\\"}',
```
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply packages/@cdktf/hcl2cdk/test/__snapshots__/variables.test.ts.snap]: There is no plan to do so since we have a self-reference here. This means var.image_id can never be a TS variable, therefore we would leave this untouched.

File: packages/@cdktf/hcl2json/lib/syntax-tree.ts (hunk 1 above)
Comment (DanielMSchmidt) [reply packages/@cdktf/hcl2json/lib/syntax-tree.ts]: forgot to remove the comment

File: packages/@cdktf/hcl2cdk/test/__snapshots__/variables.test.ts.snap (hunk 3 above)
[the remark this answers is not in the document]
Comment (ansgarm) [reply packages/@cdktf/hcl2cdk/test/__snapshots__/variables.test.ts.snap]: Alright! But couldn't it be a `ref("var.image_id")`?

[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply packages/@cdktf/hcl2cdk/test/__snapshots__/variables.test.ts.snap]: It could, but I don't see too much sense in doing this on a one off basis just for this specific edge-case. It adds another code path to our conversion code making it harder to maintain.
